File: CCKP_github/src/readdata.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class readdata {
	int n;
	double[] K;
	double[] b;
	double[][] c, a;
	int num;
	double[] zstar, zL;
	double[] timeZstar, timeZL;
	
	public void readdata(String file, int n){
		String defaultFile = "data/cckp_test_set_" + n + ".csv";
		String csvFile = (file == null || file.isBlank()) ? defaultFile : file;
        String line;
        String csvSplitBy = ",";

        int numRows = (n <= 100) ? 100 : 10;
        c = new double[numRows][];
        a = new double[numRows][];
        b = new double[numRows];
        K = new double[numRows];
        zstar = new double[numRows];
        zL = new double[numRows];
        timeZstar = new double[numRows];
        timeZL = new double[numRows];
        
        this.n = n;
        
        try (BufferedReader br = new BufferedReader(new FileReader(csvFile))) {
            int rowIndex = 0;

            while ((line = br.readLine()) != null && rowIndex < numRows) {
                
                String[] fields = line.split(csvSplitBy);
                int expectedCols = 2 * n + 6; // c[n], a[n], b, K, zstar, zL, timeZstar, timeZL
                if (fields.length < expectedCols) {
                	throw new IllegalArgumentException(
                			"Invalid CSV format in " + csvFile + " at row " + (rowIndex + 1)
                					+ ": expected at least " + expectedCols + " columns, got " + fields.length);
                }
                
                c[rowIndex] = new double[n];
                a[rowIndex] = new double[n];
                
                int loc = 0;
                for (int i = 0; i < n; i++) {
                    c[rowIndex][i] = Integer.parseInt(fields[loc]);
                    loc++;
                }
                for (int i = 0; i < n; i++) {
                    a[rowIndex][i] = Integer.parseInt(fields[loc]);
                    loc++;
                }
                b[rowIndex] = Double.parseDouble(fields[loc]);
                loc ++;
                K[rowIndex] = Double.parseDouble(fields[loc]);
                loc ++;
                zstar[rowIndex] = Double.parseDouble(fields[loc]);
                loc ++;
                zL[rowIndex] = Double.parseDouble(fields[loc]);
                loc ++;

                // Time columns are mandatory in this repository's dataset format.
                timeZstar[rowIndex] = Double.parseDouble(fields[loc]);
                loc++;
                timeZL[rowIndex] = Double.parseDouble(fields[loc]);
                loc++;
                
                rowIndex++;
            }
        } catch (IOException e) {
        	throw new RuntimeException("Failed to read CSV: " + csvFile, e);
        }
	}

	public int getn() {
		return n;
	}
	public double[][] getc(){
		return c;
	}
	public double[][] geta(){
		return a;
	}
	public double[] getb() {
		return b;
	}
	public double[] getK() {
		return K;
	}
	public double[] getzstar() {
		return zstar;
	}
	public double[] getzL() {
		return zL;
	}
	public double[] getTimeZstar() {
		return timeZstar;
	}
	public double[] getTimeZL() {
		return timeZL;
	}
	}

File: CCKP_python/src/readdata.py
"""CSV loader for CCKP instances."""

from __future__ import annotations

from dataclasses import dataclass
from typing import List


@dataclass
class CCKPData:
    n: int
    K: List[float]
    b: List[float]
    c: List[List[float]]
    a: List[List[float]]
    zstar: List[float]
    zL: List[float]
    time_zstar: List[float]
    time_zL: List[float]


def readdata(file: str | None, n: int) -> CCKPData:
    csv_file = file if file and file.strip() else f"data/cckp_test_set_{n}.csv"
    num_rows = 100 if n <= 100 else 10
    expected_cols = 2 * n + 6  # c[n], a[n], b, K, zstar, zL, timeZstar, timeZL

    c: List[List[float]] = []
    a: List[List[float]] = []
    b: List[float] = []
    K: List[float] = []
    zstar: List[float] = []
    zL: List[float] = []
    time_zstar: List[float] = []
    time_zL: List[float] = []

    with open(csv_file, "r", encoding="utf-8") as f:
        for row_index, line in enumerate(f):
            if row_index >= num_rows:
                break
            fields = line.strip().split(",")
            if len(fields) < expected_cols:
                raise ValueError(
                    f"Invalid CSV format in {csv_file} at row {row_index + 1}: "
                    f"expected at least {expected_cols} columns, got {len(fields)}"
                )

            loc = 0
            c_row = [float(int(fields[loc + i])) for i in range(n)]
            loc += n
            a_row = [float(int(fields[loc + i])) for i in range(n)]
            loc += n

            c.append(c_row)
            a.append(a_row)
            b.append(float(fields[loc]))
            loc += 1
            K.append(float(fields[loc]))
            loc += 1
            zstar.append(float(fields[loc]))
            loc += 1
            zL.append(float(fields[loc]))
            loc += 1
            time_zstar.append(float(fields[loc]))
            loc += 1
            time_zL.append(float(fields[loc]))

    return CCKPData(
        n=n,
        K=K,
        b=b,
        c=c,
        a=a,
        zstar=zstar,
        zL=zL,
        time_zstar=time_zstar,
        time_zL=time_zL,
    )
